propagators/awsxray/awsxray_propagator.go

```diff
@@ -0,0 +1,181 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aws
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	traceHeaderKey       = "X-Amzn-Trace-Id"
+	traceHeaderDelimiter = ";"
+	kvDelimiter          = "="
+	traceIDKey           = "Root"
+	sampleFlagKey        = "Sampled"
+	parentIDKey          = "Parent"
+	traceIDVersion       = "1"
+	traceIDDelimiter     = "-"
+	isSampled            = "1"
+	notSampled           = "0"
+
+	traceFlagNone           = 0x0
+	traceFlagSampled        = 0x1 << 0
+	traceIDLength           = 35
+	traceIDDelimitterIndex1 = 1
+	traceIDDelimitterIndex2 = 10
+	traceIDFirstPartLength  = 8
+	sampledFlagLength       = 1
+)
+
+var (
+	empty                  = trace.SpanContext{}
+	errInvalidTraceHeader  = errors.New("invalid X-Amzn-Trace-Id header value, should contain 3 different part separated by ;")
+	errMalformedTraceID    = errors.New("cannot decode trace id from header, should be a string of hex, lowercase trace id can't be all zero")
+	errInvalidSpanIDLength = errors.New("invalid span id length, must be 16")
+)
+
+// AWS X-Ray propagator serializes Span Context to/from AWS X-Ray headers
+//
+// AWS X-Ray format
+//
+// X-Amzn-Trace-Id: Root={traceId};Parent={parentId};Sampled={samplingFlag}
+type Xray struct{}
+
+// Asserts that the propagator implements the otel.textMapPropagator interface
+var _ propagation.TextMapPropagator = &Xray{}
+
+// Inject injects a context to the carrier following AWS X-Ray format.
+func (awsxray Xray) Inject(ctx context.Context, carrier propagation.TextMapCarrier) {
+	sc := trace.SpanFromContext(ctx).SpanContext()
+	headers := []string{}
+	if !sc.TraceID.IsValid() || !sc.SpanID.IsValid() {
+		return
+	}
+	otTraceID := sc.TraceID.String()
+	xrayTraceID := traceIDVersion + traceIDDelimiter + otTraceID[0:traceIDFirstPartLength] +
+		traceIDDelimiter + otTraceID[traceIDFirstPartLength:]
+	parentID := sc.SpanID
+	samplingFlag := notSampled
+	if sc.TraceFlags == traceFlagSampled {
+		samplingFlag = isSampled
+	}
+
+	headers = append(headers, traceIDKey, kvDelimiter, xrayTraceID, traceHeaderDelimiter, parentIDKey,
+		kvDelimiter, parentID.String(), traceHeaderDelimiter, sampleFlagKey, kvDelimiter, samplingFlag)
+
+	carrier.Set(traceHeaderKey, strings.Join(headers, ""))
+}
+
+// Extract gets a context from the carrier if it contains AWS X-Ray headers.
+func (awsxray Xray) Extract(ctx context.Context, carrier propagation.TextMapCarrier) context.Context {
+	// extract tracing information
+	if header := carrier.Get(traceHeaderKey); header != "" {
+		sc, err := extract(header)
```

Comment:
Should this error be reported with the `[ErrorHandler`](https://pkg.go.dev/go.opentelemetry.io/otel#Handle) if it is not nil?

Reply:
In the existing Jaegar and B3 propagator implementations : [B3](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/master/propagators/b3/b3_propagator.go#L154) and [Jaegar](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/master/propagators/jaeger/jaeger_propagator.go#L88) the error was also left hanging and just returns the same context as was inputted without addressing if the error was not nil. I just followed the convention of the existing implementations for the other propagators.

Reply:
We should probably update that behavior overall, but seems reasonable to be consistent. :+1: